Added SSL encryption support (basic, no verification)

alier350 opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

Added support for basic SSL encryption (with no client or server verification) via parameter `ssl_required` as an extra option - This will allow the `cassandra_keyspace` and `cassandra_role` modules to support basic SSL encryption in transit without mandatory client/server verification. This can be further extended in the future to add verification options if desired.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
- Added basic SSL support for `cassandra_keyspace` module
- Added basic SSL support for `cassandra_role` module

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
- Added support for basic SSL encryption for keyspace and role modules, to stop Ansible failing on an SSL enabled Cassandra cluster.
<!--- Paste verbatim command output below, e.g. before and after your change -->
Example Error:
```
'Error connecting to cluster: (''Unable to connect to any servers'', {''<IP>:9042'': ConnectionResetError(104, ''Connection reset by peer'')})'
```

rhysmeister commented:
Hello @alier350,

- Should also go into cassandra_table (I think that's all the non-nodetool modules).
- ssl_required should be trimmed down to ssl. This is in keeping with other db type modules.
- The cassandra_cqlsh tests fail because of cassandra_keyspace - "ImportError: cannot import name PROTOCOL_TLS". See https://github.com/ansible-collections/community.cassandra/actions/runs/3282952037/jobs/5409310749#step:7:4102. Ideally missing pylibs should be handled gracefully. See how we handle that here https://github.com/ansible-collections/community.cassandra/blob/master/plugins/modules/cassandra_table.py#L387
- This will need integration tests added to ensure it works. Ideally tests against a non-ssl enabled cluster (ssl = true and false) and on an ssl enabled cluster /ssl = true and false). This will ensure the module handles everything gracefully.
- Last 2 lint issues are...

> 
> ERROR: plugins/modules/cassandra_keyspace.py:0:0: doc-default-does-not-match-spec: Argument 'ssl_required' in argument_spec defines default as (False) but documentation defines default as (None)
> ERROR: plugins/modules/cassandra_role.py:0:0: doc-default-does-not-match-spec: Argument 'ssl_required' in argument_spec defines default as (False) but documentation defines default as (None)

alier350 commented:
Hello Rhys, 

Thank you for your feedback, very useful, I will look into implementing the above improvements and get back to you.

Best,
Alex

alier350 commented:
As discussed, I have:
- Added `ssl` in cassandra_table
- ssl_required trimmed down to ssl on all modules
- handled SSL lib import gracefully
- Fixed remaining linter issues

Integration tests, had a look on https://github.com/riptano/ccm but after spinning up a test cluster, it is not ssl enabled and it doesn't look straightforward to so from the options, I'll see if there is way to re-use the existing tests but this might take longer to work on.

rhysmeister commented:
Creating an ssl enabled cluster should be possible with ccm. I'm unable to find a full example atm. The following links should be useful...

https://github.com/riptano/ccm/issues/179
https://github.com/riptano/ccm/issues?q=is%3Aissue+ssl

I might have a stab when I get some time.

rhysmeister commented on plugins/modules/cassandra_keyspace.py:
```diff
@@ -250,10 +261,17 @@ def main():
                " install cassandra-driver.")
         module.fail_json(msg=msg)
 
+    if HAS_SSL_LIBRARY is False:
```
Let's also check if the ssl parameter is true here as well. That way we only require the ssl lib when it's actually used.

alier350 replied:
Yep, done

rhysmeister commented on plugins/modules/cassandra_role.py:
```diff
@@ -556,8 +567,15 @@ def main():
                " install cassandra-driver.")
         module.fail_json(msg=msg)
 
+    if HAS_SSL_LIBRARY is False:
```
Ditto

rhysmeister commented on plugins/modules/cassandra_table.py:
```diff
@@ -390,8 +401,15 @@ def main():
                " install cassandra-driver.")
         module.fail_json(msg=msg)
 
+    if HAS_SSL_LIBRARY is False:
```
Ditto.

rhysmeister commented:
For the integration tests I would pick one module for now, i.e. cassandra_table, then...

1. Test without the SSL driver installed (use ignore_errors and verifiy the output).
2. Install the SSL driver.
3. Test ssl = yes to a non ssl enabled cluster - how is this case handled.

Once that's done we can worry about the SSL-Enabled Cluster.

alier350 commented:
Added a check condition for ssl=True on a non-ssl cluster - From my local testing it should fail, and I see this error from the module 
```
'Error connecting to cluster: (''Unable to connect to any servers'', {''172.21.30.36:9042'': PermissionError(1, "Tried connecting to [(''<IP>'', 9042)]. Last error: [SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:1131)")})'
```
For SSL, there is no driver exactly, it's an internal module since Python 2.6 and over from what I gather at https://wiki.python.org/moin/SSL, however we could still add a check for `pip install ssl` to cover the lower Python versions?

rhysmeister commented:
> Added a check condition for ssl=True on a non-ssl cluster - From my local testing it should fail, and I see this error from the module
> 
> ```
> 'Error connecting to cluster: (''Unable to connect to any servers'', {''172.21.30.36:9042'': PermissionError(1, "Tried connecting to [(''<IP>'', 9042)]. Last error: [SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:1131)")})'
> ```
> 
> For SSL, there is no driver exactly, it's an internal module since Python 2.6 and over from what I gather at https://wiki.python.org/moin/SSL, however I could still add a check for `pip install ssl` to cover the lower Python versions?

I'm not sure that can be the case. I saw the module throwing this error. We don't test on anything lower than 2.7. I'm pretty sure I never have for this collection.

rhysmeister commented:
OK. Soemthing funny is going on...

```
TASK [cassandra_table : Check keyspace creation on a non-SSL cluster] **********
fatal: [testhost]: FAILED! => {"changed": false, "msg": "This module requires the SSL python library. You can probably install it with pip install ssl."}
...ignoring

TASK [cassandra_table : assert] ************************************************
fatal: [testhost]: FAILED! => {
    "assertion": "'SSL: WRONG_VERSION_NUMBER' in nonssl_keyspace.msg",
    "changed": false,
    "evaluated_to": false,
    "msg": "Assertion failed"
}
```

So either the logic is wrong here or the ssl really isn't included. If ssl isn't there, this case can be tested for, then ssl installed. We don't run on anything under Python 2.7.

sanity test dev - You can ignore this one. This is a new test introduced today and that's been fixed and merged to master already.

alier350 commented:
Hm, it is peculiar. I tested locally on a fresh Ubuntu 20.04 container installing Python 2.7 ans the ssl module is indeed included. I added an extra step in the last commit to `pip install ssl` just in case but the tests failed with `ValueError: This extension should not be used with Python 2.6 or later (already built in), and has not been tested with Python 2.3.4 or earlier` in https://github.com/ansible-collections/community.cassandra/actions/runs/3384529823/jobs/5622337449#step:7:3028

I'll have another look at the logic

rhysmeister commented:
Perhaps try /usr/bin/python here...

>  fatal: [testhost]: FAILED! => {
>     "changed": true,
>     "cmd": "python -c \"import ssl\"",
>     "delta": "0:00:00.004895",
>     "end": "2022-11-09 17:50:33.696323",
>     "invocation": {
>         "module_args": {
>             "_raw_params": "python -c \"import ssl\"",
>             "_uses_shell": true,
>             "argv": null,
>             "chdir": null,
>             "creates": null,
>             "executable": null,
>             "removes": null,
>             "stdin": null,
>             "stdin_add_newline": true,
>             "strip_empty_ends": true
>         }
>     },
>     "msg": "non-zero return code",
>     "rc": 127,
>     "start": "2022-11-09 17:50:33.691428",
>     "stderr": "/bin/sh: 1: python: not found",
>     "stderr_lines": [
>         "/bin/sh: 1: python: not found"
>     ],
>     "stdout": "",
>     "stdout_lines": []
> }

alier350 commented:
Added SSL lib checks and assertions on the other 2 modules as well (`cassandra_keyspace` and `cassandra_role`) as the tests looked initially successful. I do see a new error on the latest tests, though I'm not sure it is related to the new SSL tests:
```
msg": "Failed to download key at https://www.apache.org/dist/cassandra/KEYS: Request failed: <urlopen error [Errno 99] Cannot assign requested address>"
```

rhysmeister commented:
> Added SSL lib checks and assertions on the other 2 modules as well (`cassandra_keyspace` and `cassandra_role`) as the tests looked initially successful. I do see a new error on the latest tests, though I'm not sure it is related to the new SSL tests:
> 
> ```
> msg": "Failed to download key at https://www.apache.org/dist/cassandra/KEYS: Request failed: <urlopen error [Errno 99] Cannot assign requested address>"
> ```

Probably just a network clitch here.

rhysmeister commented on tests/integration/targets/cassandra_role/tasks/main.yml:
```diff
@@ -117,6 +140,26 @@
+  ansible.builtin.shell: cqlsh --username "{{ cassandra_admin_user }}" --password "{{ cassandra_admin_pwd }}" --execute "LIST ROLES"
+  register: nonssl_myrole
+
+- name: Assert no_ssl_app_user exists in output
```
You're using check_mode: yes above so this role shouldn't exist and this assert will always fail.

alier350 replied:
Yep, fixed in https://github.com/ansible-collections/community.cassandra/pull/230/commits/5ef4e7b36719cd8b323b10445189338685016e2a

rhysmeister commented:
OK the default verify_mode seems to be VerifyMode.CERT_NONE so that's why the create works...

> >>> from ssl import SSLContext, PROTOCOL_TLS
> >>> ssl_context = SSLContext(PROTOCOL_TLS)
> >>> ssl_context.verify_mode
> <VerifyMode.CERT_NONE: 0>

Perhaps it's sensible to set this to a secure default. We do this in the mongodb collection with a module parameter as follows...

https://github.com/ansible-collections/community.mongodb/blob/master/plugins/module_utils/mongodb_common.py

Valid values...

CERT_REQUIRED
CERT_OPTIONAL
CERT_NONE

Documentation...

https://pythontic.com/ssl/sslcontext/verify_mode

Rhys

alier350 commented:
> OK the default verify_mode seems to be VerifyMode.CERT_NONE so that's why the create works...
> 
> > > > > from ssl import SSLContext, PROTOCOL_TLS
> > > > > ssl_context = SSLContext(PROTOCOL_TLS)
> > > > > ssl_context.verify_mode
> > > > > <VerifyMode.CERT_NONE: 0>
> 
> Perhaps it's sensible to set this to a secure default. We do this in the mongodb collection with a module parameter as follows...
> 
> https://github.com/ansible-collections/community.mongodb/blob/master/plugins/module_utils/mongodb_common.py
> 
> Valid values...
> 
> CERT_REQUIRED CERT_OPTIONAL CERT_NONE
> 
> Documentation...
> 
> https://pythontic.com/ssl/sslcontext/verify_mode
> 
> Rhys
> 
> UPDATE: I've just added this.... thinking I might as well add a parameter for a cert as well even if just for future needs.

Yep, the ssl_context is much better defined now, thanks! Next I'll try and see if I can get integration tests with SSL enabled and the ccm utility working. Though current tests appear to fail on authentication assertion for `TASK [cassandra_role : Create a role - Auth is not enabled so will fail]` https://github.com/ansible-collections/community.cassandra/actions/runs/3559341210/jobs/5978679377#step:7:3393, I haven't made any changes there, maybe a blip.

rhysmeister commented:
I don't think the change is correct...

https://github.com/ansible-collections/community.cassandra/actions/runs/3559341210/jobs/5978679377#step:7:579

I would expect this to fail as SSL is not enabled on the cluster. I've not investigated why yet.

eugenetolan commented:
Any progress on this @rhysmeister?

rhysmeister commented:
This won't be on my radar anytime soon. Community fixes appreciated!

R

JakubBiel commented on tests/integration/targets/cassandra_role/tasks/main.yml:
```diff
@@ -117,6 +140,25 @@
+    password: 'secretZHB78'
+    state: present
+    login: yes
+    ssl: False
```
Shouldn't this be True?

alier350 replied:
That particular test is set to ensure a role is created on a non-SSL cluster when we pass `ssl: False` - Although it is a good point, probably another test can be added when we set `ssl: True` and that would be expected to fail.

alier350 commented:
Just to drop a quick update, with the last commit in November checks are passing but we still see a failure due to `codecov/project` reporting reduced coverage (not fully sure why, since we've been adding tests). 

I wasn't able to follow up on this recently due to some other work, but what is mostly missing are integration tests for SSL enabled clusters which I'm trying to add using the CCM utility (https://github.com/riptano/ccm) as we discussed above with @rhysmeister . That has proven to be a bit more fiddly to set up and test but contributions are welcome. I'll try and commit some initial tests I've been working on at some point later.

JakubBiel commented:
Hey @rhysmeister What are your thoughts about new tests that would run TLS-enabled clusters? Should this go into a separate workflow, for example `ansible-test-ssl.yml` and then run all the scenarios against 3 modules, i.e. cassandra_keyspace, cassandra_table and cassandra_role. That would give us around 30 tests.

rhysmeister commented:
Tests are currently contained per module...

https://github.com/ansible-collections/community.cassandra/tree/master/tests/integration/targets

If it makes sense to do it a different way then I'm open to that.

alier350 commented:
Added a `cassandra_ssl` directory to run tests with CCM - I copied some of the logic from `setup_cassandra_cluster_manager` for the CCM tasks, I can review if this causes too much duplication

rhysmeister commented:
> Added a `cassandra_ssl` directory to run tests with CCM - I copied some of the logic from `setup_cassandra_cluster_manager` for the CCM tasks, I can review if this causes too much duplication

Currently throwing the following error...

> TASK [setup_cassandra : Setup Cassandra apt repo] ******************************
> fatal: [testhost]: FAILED! => {"attempts": 3, "changed": false, "msg": "Failed to update apt cache: E:The repository 'http://www.apache.org/dist/cassandra/debian 3.11.14 Release' does not have a Release file."}

This needs to be set to one of the following in...

https://downloads.apache.org/cassandra/debian/dists/

i.e. 40x for the apt repo to be added. I think later on it might need to be overridden for CCM. I forget how this works exactly but you're probably better off running the tests locally.

alier350 commented:
Hi @rhysmeister - Not sure, but there seems to be an issue with the latest tests that appear to be unrelated with our changes, I see failures in:
- [integration_311x (3.9, devel, 311x, 3.11.11, centos7, 2.7, --docker-privileged, cassandra_flush)](https://github.com/ansible-collections/community.cassandra/actions/runs/4594297753/jobs/8113993527?pr=230#logs) `Could not detect which major revision of dnf is in use, which is required to determine module backend.`
- [integration_40x (3.9, devel, 40x, 4.0.5, ubuntu2004, 3.8, cassandra_garbagecollect)](https://github.com/ansible-collections/community.cassandra/actions/runs/4594297753/jobs/8113956155?pr=230#logs) `"msg": "Failed to download key at https://www.apache.org/dist/cassandra/KEYS: Request failed: <urlopen error [Errno 99] Cannot assign requested address>"`

Maybe we can re-trigger the tests? I wouldn't expect our CCM changes to affect anything other than our `cassandra_ssl` test module, but could be wrong :)

rhysmeister commented:
> Hi @rhysmeister - Not sure, but there seems to be an issue with the latest tests that appear to be unrelated with our changes, I see failures in:
> 
>     * [integration_311x (3.9, devel, 311x, 3.11.11, centos7, 2.7, --docker-privileged, cassandra_flush)](https://github.com/ansible-collections/community.cassandra/actions/runs/4594297753/jobs/8113993527?pr=230#logs) `Could not detect which major revision of dnf is in use, which is required to determine module backend.`
> 
>     * [integration_40x (3.9, devel, 40x, 4.0.5, ubuntu2004, 3.8, cassandra_garbagecollect)](https://github.com/ansible-collections/community.cassandra/actions/runs/4594297753/jobs/8113956155?pr=230#logs) `"msg": "Failed to download key at https://www.apache.org/dist/cassandra/KEYS: Request failed: <urlopen error [Errno 99] Cannot assign requested address>"`
> 
> 
> Maybe we can re-trigger the tests? I wouldn't expect our CCM changes to affect anything other than our `cassandra_ssl` test module, but could be wrong :)

Your're not wrong. 
- cassandra_flush - Probably an upstream change that we will hit again.
- cassandra_garbagecollect - Network glich.

You'll be better of running these locally as the dnf problem will probably happen again until fix (either by me of upstream).

alier350 commented:
Hi @rhysmeister , was wondering if you've been able to fix the integration tests, I've been trying to test this locally with ACT (https://github.com/nektos/act) but with not much success, the workflows fail when attempting to run the sanity and integration tests with just `[CI/sanity-2] exitcode '1': failure`. 
In your experience, is there any other way we can run this tests locally or some tool you can recommend to get these tests running in a local box?

rhysmeister commented:
Hi @alier350,

There are some instructions on how to run the integration tests locally in the [README.md](https://github.com/ansible-collections/community.cassandra/blob/master/README.md file. I do this on MacOS and Linux with no issues.

CHeers,

Rhys

rhysmeister commented:
Tests for Cassandra 4 are failing. 

[integration_40x (3.9, devel, 40x, 4.0.5, ubuntu2004, 3.8, cassandra_ssl)](https://github.com/ansible-collections/community.cassandra/actions/runs/5222629946/jobs/9442834241?pr=230#logs)

The error is...

> "ERROR [main] 2023-06-10 11:40:33,464 CassandraDaemon.java:911 - Cannot start node if snitch's data center (DC1) differs from previous data center (datacenter1). Please fix the snitch configuration, decommission and rebootstrap this node or use the flag -Dcassandra.ignore_dc=true.“

This must be a new check in 4 that 3.11 didn't do. 

In setup_cassandra we should also explicitly fail when this happens. At the moment we do...

setup_cassandra/tasks/main.yml

```
- name: End play if wait_for task failed
  meta: end_play
  when:
    - cassandra_ports.failed is defined
    - cassandra_ports.failed
```

and this should probably be something like this so the failure is spotted..

- name: End play if wait_for task failed
  fail:
    msg: "Cassandra failed to start"
  when:
    - cassandra_ports.failed is defined
    - cassandra_ports.failed

I've just opened a new PR to include a static KEYS file instead of always calling it from the url. This will fix a lot of the annoying timeouts we've seen in these tests. You might want ot rebase on that.

https://github.com/ansible-collections/community.cassandra/pull/247

rhysmeister commented:
> Tests for Cassandra 4 are failing.
> 
> [integration_40x (3.9, devel, 40x, 4.0.5, ubuntu2004, 3.8, cassandra_ssl)](https://github.com/ansible-collections/community.cassandra/actions/runs/5222629946/jobs/9442834241?pr=230#logs)
> 
> The error is...
> 
> > "ERROR [main] 2023-06-10 11:40:33,464 CassandraDaemon.java:911 - Cannot start node if snitch's data center (DC1) differs from previous data center (datacenter1). Please fix the snitch configuration, decommission and rebootstrap this node or use the flag -Dcassandra.ignore_dc=true.“
> 
> This must be a new check in 4 that 3.11 didn't do.
> 
> In setup_cassandra we should also explicitly fail when this happens. At the moment we do...
> 
> setup_cassandra/tasks/main.yml
> 
> ```
> - name: End play if wait_for task failed
>   meta: end_play
>   when:
>     - cassandra_ports.failed is defined
>     - cassandra_ports.failed
> ```
> 
> and this should probably be something like this so the failure is spotted..
> 
> ```
> - name: End play if wait_for task failed
>   fail:
>     msg: "Cassandra failed to start"
>   when:
>     - cassandra_ports.failed is defined
>     - cassandra_ports.failed
> ```
> 
> I've just opened a new PR to include a static KEYS file instead of always calling it from the url. This will fix a lot of the annoying timeouts we've seen in these tests. You might want ot rebase on that.
> 
> #247

I've added a few commits that should address this.

alier350 commented:
Got it, thanks! - Pulled and rebased to master, hopefully it goes through

Sidenote - The integration tests for SSL modules are nearly complete, I need to add a couple more test cases to cover but we should be nearly there

rhysmeister commented:
> > Error from cassandra_ssl
> > ```
> > TASK [cassandra_ssl : Creating a keyspace with ssl=True and verify_mode=CERT_REQUIRED on an SSL enabled cluster should succeed] ***
> > 
> > fatal: [testhost]: FAILED! => {"changed": false, "msg": "An error occured: errors={'Connection defunct by heartbeat': 'Client request timeout. See Session.execute[_async](timeout)'}, last_host=127.0.0.5:9042"}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > https://github.com/ansible-collections/community.cassandra/actions/runs/5242825941/jobs/9470871674?pr=230
> 
> Tested again locally and it seems to work:
> 
> ```
> TASK [cassandra_ssl : Creating a keyspace with ssl=True and verify_mode=CERT_REQUIRED on an SSL enabled cluster should succeed] ***
> changed: [testhost] => {"changed": true, "cql": "CREATE KEYSPACE ccmsslkeyspace_cert_required WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor': 1 } AND DURABLE_WRITES = True", "keyspace": "ccmsslkeyspace_cert_required"}
> 
> TASK [cassandra_ssl : Get output of DESC KEYSPACE ccmsslkeyspace_cert_required] ***
> changed: [testhost] => {"changed": true, "cmd": "export SSL_CERTFILE=/home/ssl-test-cluster/ssl-test.cer.pem; cqlsh --ssl --execute \"DESC KEYSPACE ccmsslkeyspace_cert_required\"", "delta": "0:00:00.535858", "end": "2023-06-12 15:19:05.095632", "rc": 0, "start": "2023-06-12 15:19:04.559774", "stderr": "", "stderr_lines": [], "stdout": "\nCREATE KEYSPACE ccmsslkeyspace_cert_required WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '1'}  AND durable_writes = true;", "stdout_lines": ["", "CREATE KEYSPACE ccmsslkeyspace_cert_required WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '1'}  AND durable_writes = true;"]}
> 
> TASK [cassandra_ssl : Assert ccmsslkeyspace_cert_required exists 3] ************
> ok: [testhost] => {
>     "changed": false,
>     "msg": "All assertions passed"
> }
> ```
> 
> Maybe a transient timeout error?

Yes I guess so. There's another weird error happening sometimes...

> Cannot start node if snitch's data center (DC1) differs from previous data center (datacenter1). Please fix the snitch configuration, decommission and rebootstrap this node or use the flag -Dcassandra.ignore_dc=true

This seem to happen on Cassandra 4 only. I added the suggested flag to the cassandra startup script but it still seem to happen... but only sometimes . I don't know why this happens only sometimes but I don 't think it's anything to do with your stuff here. Unless you happen to know about it I'd suggest we ignore it here.

alier350 commented:
> > > Error from cassandra_ssl
> > > ```
> > > TASK [cassandra_ssl : Creating a keyspace with ssl=True and verify_mode=CERT_REQUIRED on an SSL enabled cluster should succeed] ***
> > > 
> > > fatal: [testhost]: FAILED! => {"changed": false, "msg": "An error occured: errors={'Connection defunct by heartbeat': 'Client request timeout. See Session.execute[_async](timeout)'}, last_host=127.0.0.5:9042"}
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > https://github.com/ansible-collections/community.cassandra/actions/runs/5242825941/jobs/9470871674?pr=230
> > 
> > 
> > Tested again locally and it seems to work:
> > ```
> > TASK [cassandra_ssl : Creating a keyspace with ssl=True and verify_mode=CERT_REQUIRED on an SSL enabled cluster should succeed] ***
> > changed: [testhost] => {"changed": true, "cql": "CREATE KEYSPACE ccmsslkeyspace_cert_required WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor': 1 } AND DURABLE_WRITES = True", "keyspace": "ccmsslkeyspace_cert_required"}
> > 
> > TASK [cassandra_ssl : Get output of DESC KEYSPACE ccmsslkeyspace_cert_required] ***
> > changed: [testhost] => {"changed": true, "cmd": "export SSL_CERTFILE=/home/ssl-test-cluster/ssl-test.cer.pem; cqlsh --ssl --execute \"DESC KEYSPACE ccmsslkeyspace_cert_required\"", "delta": "0:00:00.535858", "end": "2023-06-12 15:19:05.095632", "rc": 0, "start": "2023-06-12 15:19:04.559774", "stderr": "", "stderr_lines": [], "stdout": "\nCREATE KEYSPACE ccmsslkeyspace_cert_required WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '1'}  AND durable_writes = true;", "stdout_lines": ["", "CREATE KEYSPACE ccmsslkeyspace_cert_required WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '1'}  AND durable_writes = true;"]}
> > 
> > TASK [cassandra_ssl : Assert ccmsslkeyspace_cert_required exists 3] ************
> > ok: [testhost] => {
> >     "changed": false,
> >     "msg": "All assertions passed"
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Maybe a transient timeout error?
> 
> Yes I guess so. There's another weird error happening sometimes...
> 
> > Cannot start node if snitch's data center (DC1) differs from previous data center (datacenter1). Please fix the snitch configuration, decommission and rebootstrap this node or use the flag -Dcassandra.ignore_dc=true
> 
> This seem to happen on Cassandra 4 only. I added the suggested flag to the cassandra startup script but it still seem to happen... but only sometimes . I don't know why this happens only sometimes but I don 't think it's anything to do with your stuff here. Unless you happen to know about it I'd suggest we ignore it here.

Sure - Though for the SSL tests we do not really need the local Cassandra setup, all tests are done through CCM so I might end up removing the `setup_cassandra` dependency for `cassandra_ssl` to make the tests a bit more robust, I'll look into it.

rhysmeister commented:
> > > > Error from cassandra_ssl
> > > > ```
> > > > TASK [cassandra_ssl : Creating a keyspace with ssl=True and verify_mode=CERT_REQUIRED on an SSL enabled cluster should succeed] ***
> > > > 
> > > > fatal: [testhost]: FAILED! => {"changed": false, "msg": "An error occured: errors={'Connection defunct by heartbeat': 'Client request timeout. See Session.execute[_async](timeout)'}, last_host=127.0.0.5:9042"}
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > https://github.com/ansible-collections/community.cassandra/actions/runs/5242825941/jobs/9470871674?pr=230
> > > 
> > > 
> > > Tested again locally and it seems to work:
> > > ```
> > > TASK [cassandra_ssl : Creating a keyspace with ssl=True and verify_mode=CERT_REQUIRED on an SSL enabled cluster should succeed] ***
> > > changed: [testhost] => {"changed": true, "cql": "CREATE KEYSPACE ccmsslkeyspace_cert_required WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor': 1 } AND DURABLE_WRITES = True", "keyspace": "ccmsslkeyspace_cert_required"}
> > > 
> > > TASK [cassandra_ssl : Get output of DESC KEYSPACE ccmsslkeyspace_cert_required] ***
> > > changed: [testhost] => {"changed": true, "cmd": "export SSL_CERTFILE=/home/ssl-test-cluster/ssl-test.cer.pem; cqlsh --ssl --execute \"DESC KEYSPACE ccmsslkeyspace_cert_required\"", "delta": "0:00:00.535858", "end": "2023-06-12 15:19:05.095632", "rc": 0, "start": "2023-06-12 15:19:04.559774", "stderr": "", "stderr_lines": [], "stdout": "\nCREATE KEYSPACE ccmsslkeyspace_cert_required WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '1'}  AND durable_writes = true;", "stdout_lines": ["", "CREATE KEYSPACE ccmsslkeyspace_cert_required WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '1'}  AND durable_writes = true;"]}
> > > 
> > > TASK [cassandra_ssl : Assert ccmsslkeyspace_cert_required exists 3] ************
> > > ok: [testhost] => {
> > >     "changed": false,
> > >     "msg": "All assertions passed"
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Maybe a transient timeout error?
> > 
> > 
> > Yes I guess so. There's another weird error happening sometimes...
> > > Cannot start node if snitch's data center (DC1) differs from previous data center (datacenter1). Please fix the snitch configuration, decommission and rebootstrap this node or use the flag -Dcassandra.ignore_dc=true
> > 
> > 
> > This seem to happen on Cassandra 4 only. I added the suggested flag to the cassandra startup script but it still seem to happen... but only sometimes . I don't know why this happens only sometimes but I don 't think it's anything to do with your stuff here. Unless you happen to know about it I'd suggest we ignore it here.
> 
> Sure - Though for the SSL tests we do not really need the local Cassandra setup, all tests are done through CCM so I might end up removing the `setup_cassandra` dependency for `cassandra_ssl` to make the tests a bit more robust, I'll look into it.

I cannot for the life of me remember why but I think the setup_cassandra needs to be there.

rhysmeister commented:
One of example of the intermittent problem...

>"INFO [main] 2023-06-16 20:11:36,461 QueryProcessor.java:115 - Initialized prepared statement caches with 10 MB",
> "ERROR [main] 2023-06-16 20:11:36,618 CassandraDaemon.java:911 - Cannot start node if snitch's data center (DC1) differs from previous data center (datacenter1). Please fix the snitch configuration, decommission and rebootstrap this node or use the flag -Dcassandra.ignore_dc=true."

https://github.com/ansible-collections/community.cassandra/actions/runs/5288742741/jobs/9581909657?pr=230

alier350 commented:
> > > > > Error from cassandra_ssl
> > > > > ```
> > > > > TASK [cassandra_ssl : Creating a keyspace with ssl=True and verify_mode=CERT_REQUIRED on an SSL enabled cluster should succeed] ***
> > > > > 
> > > > > fatal: [testhost]: FAILED! => {"changed": false, "msg": "An error occured: errors={'Connection defunct by heartbeat': 'Client request timeout. See Session.execute[_async](timeout)'}, last_host=127.0.0.5:9042"}
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > https://github.com/ansible-collections/community.cassandra/actions/runs/5242825941/jobs/9470871674?pr=230
> > > > 
> > > > 
> > > > Tested again locally and it seems to work:
> > > > ```
> > > > TASK [cassandra_ssl : Creating a keyspace with ssl=True and verify_mode=CERT_REQUIRED on an SSL enabled cluster should succeed] ***
> > > > changed: [testhost] => {"changed": true, "cql": "CREATE KEYSPACE ccmsslkeyspace_cert_required WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor': 1 } AND DURABLE_WRITES = True", "keyspace": "ccmsslkeyspace_cert_required"}
> > > > 
> > > > TASK [cassandra_ssl : Get output of DESC KEYSPACE ccmsslkeyspace_cert_required] ***
> > > > changed: [testhost] => {"changed": true, "cmd": "export SSL_CERTFILE=/home/ssl-test-cluster/ssl-test.cer.pem; cqlsh --ssl --execute \"DESC KEYSPACE ccmsslkeyspace_cert_required\"", "delta": "0:00:00.535858", "end": "2023-06-12 15:19:05.095632", "rc": 0, "start": "2023-06-12 15:19:04.559774", "stderr": "", "stderr_lines": [], "stdout": "\nCREATE KEYSPACE ccmsslkeyspace_cert_required WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '1'}  AND durable_writes = true;", "stdout_lines": ["", "CREATE KEYSPACE ccmsslkeyspace_cert_required WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '1'}  AND durable_writes = true;"]}
> > > > 
> > > > TASK [cassandra_ssl : Assert ccmsslkeyspace_cert_required exists 3] ************
> > > > ok: [testhost] => {
> > > >     "changed": false,
> > > >     "msg": "All assertions passed"
> > > > }
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > Maybe a transient timeout error?
> > > 
> > > 
> > > Yes I guess so. There's another weird error happening sometimes...
> > > > Cannot start node if snitch's data center (DC1) differs from previous data center (datacenter1). Please fix the snitch configuration, decommission and rebootstrap this node or use the flag -Dcassandra.ignore_dc=true
> > > 
> > > 
> > > This seem to happen on Cassandra 4 only. I added the suggested flag to the cassandra startup script but it still seem to happen... but only sometimes . I don't know why this happens only sometimes but I don 't think it's anything to do with your stuff here. Unless you happen to know about it I'd suggest we ignore it here.
> > 
> > 
> > Sure - Though for the SSL tests we do not really need the local Cassandra setup, all tests are done through CCM so I might end up removing the `setup_cassandra` dependency for `cassandra_ssl` to make the tests a bit more robust, I'll look into it.
> 
> I cannot for the life of me remember why but I think the setup_cassandra needs to be there.

It kinda does need to be there, yeah, it installs cqlsh and Java which are then used by the CCM playbook. I did however manage to work around it on my latest commit, by having the CCM playbook installing the Java JDK if java is not there and having the ssl playbook install cqlsh itself, with my local tests passing. 

This should save some time when running the SSL tests and avoiding some instances where the snitch DC error might occure, though admiteddly it doesn't help other modules much. 

Let me know what you think, happy to revert if you prefer to have all module tests with `setup_cassandra` included

alier350 commented:
Hm, cqlsh 6.1.0 didn't play nice with older Cassandra versions:

```DESCRIBE|DESC was moved to server side in Cassandra 4.0. As a consequence DESRIBE|DESC will not work in cqlsh '6.1.0' connected to Cassandra '3.11.11', the version that you are connected to. DESCRIBE does not exist server side prior Cassandra 4.0.", "stdout_lines": ["", "WARN: DESCRIBE|DESC was moved to server side in Cassandra 4.0. As a consequence DESRIBE|DESC will not work in cqlsh '6.1.0' connected to Cassandra '3.11.11', the version that you are connected to. DESCRIBE does not exist server side prior Cassandra 4.0."]}```

I've reverted the previous commit to always run the Cassandra setup for the cassandra_ssl tests as it was before, it wasn't that crucial - it would be nice to have to improve on the running times for tests but not worth breaking older Cassandra version tests

alier350 commented:
Last thing, I was having a look at the coverage failure report (`codecov/project — 77.58% (-0.28%) compared to 0c4556c `), which is the last test failing, though not sure what to fix there, trying to dropdown the report just gives a blank line:
![Screenshot 2023-06-22 at 12 39 45](https://github.com/ansible-collections/community.cassandra/assets/21689354/e6843be0-9782-4a67-9124-81fe7bb53e68)

Also, there's mention of `cassandra_reload.py` coverage, which i didn't touch, let me know if there's something I can do there @rhysmeister

rhysmeister commented:
@alier350,

Let's ignore codecov for the moment, 0.28% is pretty trivial anyway. Are you done here? I will hopefully get some time this weekend to do a full proper review. It's been so long since this was opened and I think it deserved a final review before merge.

Cheers,

Rhys

alier350 commented:
> @alier350,
> 
> Let's ignore codecov for the moment, 0.28% is pretty trivial anyway. Are you done here? I will hopefully get some time this weekend to do a full proper review. It's been so long since this was opened and I think it deserved a final review before merge.
> 
> Cheers,
> 
> Rhys

Yep, I'm done, please re-review since as you said it's been some time and needs another look now - In terms of what's been changed recently we have:
- ssl_verify_location: This is the SSL CA chain or certificate path the driver needs to load to confirm if the certificate presented to it by a Cassandra server is valid or not when verify_mode is set to CERT_OPTIONAL or CERT_REQUIRED (docs https://docs.datastax.com/en/developer/python-driver/3.25/security/#ssl-configuration-examples)
- Added sanity checks to ensure that when verify_mode is set to CERT_OPTIONAL or CERT_REQUIRED an ssl_verify location is also supplied otherwise we fail the module with the appropriate error (`ssl_verify_location is also required to be set and not empty` or `ssl_verify_location certificate: File not found`) - For example https://github.com/alier350/community.cassandra/blob/cassandra_roles_keyspaces_ssl_support/plugins/modules/cassandra_role.py#L613-L620
- Integration tests - Added the cassandra_ssl tests in its entirety and added tests for all modules using SSL, roles, keyspaces and tables trying to cover all positive and negative cases I could think - I've also signed and created new certificates and keystores/trustores for CCM to use when spinning up a test cluster in https://github.com/alier350/community.cassandra/tree/cassandra_roles_keyspaces_ssl_support/tests/integration/targets/cassandra_ssl/files, the ones provided in the CCM repo had expired in 2018

rhysmeister commented on plugins/modules/cassandra_keyspace.py:
Can we rename these varibales

verify_mode -> ssl_cert_reqs
ssl_verify_location ->  ssl_ca_certs

I think it makes sense to normalise these, for example the mongodb modules use these names... https://github.com/ansible-collections/community.mongodb/blob/master/plugins/doc_fragments/ssl_options.py

alier350 replied:
Done in [b79fd7b](https://github.com/ansible-collections/community.cassandra/pull/230/commits/b79fd7beb470191d7b96b3ba64a0a3b34b3cbc2b)

rhysmeister commented on plugins/modules/cassandra_keyspace.py:
```diff
@@ -268,6 +302,24 @@ def main():
+    verify_mode = module.params['verify_mode']
+    ssl_verify_location = module.params['ssl_verify_location']
+
+    if verify_mode in ('CERT_REQUIRED', 'CERT_OPTIONAL') and module.params['ssl_verify_location'] == '':
```
Could shorten module.params['ssl_verify_location'] to ssl_verify_location

alier350 replied:
Also done in [b79fd7b](https://github.com/ansible-collections/community.cassandra/pull/230/commits/b79fd7beb470191d7b96b3ba64a0a3b34b3cbc2b)

rhysmeister commented on tests/integration/targets/cassandra_ssl/tasks/main.yml:
Looks liek a comprehensive set of tests

rhysmeister reviewed (changes requested):
Looks good.

I think there's some code that should be refactored into module_utils (new code as well as a lot of existing code). For example...

- shared plugin parameters
- plugin documentation
- a bit of validation code

I already do this for the nodetool based plugins. I'll create another issue for that so let's nt worry about that here. That can be done be myself, you or someone else completely.

Well done for persisting with this!

Rhys

alier350 commented:
> 

Thanks! That's good, I'll look into refactoring those variables when I get some time available this week, shouldn't take too long.

alier350 commented:
Great stuff, thanks Rhys! And thank you for bearing with us for the last few months :)

rhysmeister commented:
No sweat. As previously said, I'm impressed you stuck with it.

mks-sky commented:
Any idea when this might make it into v1.2.5 of https://galaxy.ansible.com/community/cassandra?  We're hoping to make use of this SSL encryption support.
